ports/nrf/common-hal/alarm/__init__.c

```diff
@@ -0,0 +1,399 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/gc.h"
+#include "py/obj.h"
+#include "py/objtuple.h"
+#include "py/runtime.h"
+#include <stdio.h>
+#include <string.h>
+
+#include "shared-bindings/alarm/__init__.h"
+#include "shared-bindings/alarm/SleepMemory.h"
+#include "shared-bindings/alarm/pin/PinAlarm.h"
+#include "shared-bindings/alarm/time/TimeAlarm.h"
+#include "shared-bindings/alarm/touch/TouchAlarm.h"
+#include "shared-bindings/time/__init__.h"
+
+#include "supervisor/port.h"
+#include "supervisor/serial.h"  // serial_connected()
+#ifdef NRF_DEBUG_PRINT
+#include "supervisor/serial.h" // dbg_printf()
+extern int dbg_check_RTCprescaler(void);
+#endif
+#include "supervisor/qspi_flash.h"
+
+#include "nrf.h"
+#include "nrf_power.h"
+#include "nrfx.h"
+#include "nrfx_gpiote.h"
+
+// Singleton instance of SleepMemory.
+const alarm_sleep_memory_obj_t alarm_sleep_memory_obj = {
+    .base = {
+        .type = &alarm_sleep_memory_type,
+    },
+};
+
+void alarm_reset(void) {
+    alarm_sleep_memory_reset();
+    alarm_pin_pinalarm_reset();
+    alarm_time_timealarm_reset();
+    alarm_touch_touchalarm_reset();
+}
+
+extern uint32_t reset_reason_saved;
+STATIC nrf_sleep_source_t _get_wakeup_cause(void) {
+    if (alarm_pin_pinalarm_woke_us_up()) {
+        return NRF_SLEEP_WAKEUP_GPIO;
+    }
+    if (alarm_time_timealarm_woke_us_up()) {
+        return NRF_SLEEP_WAKEUP_TIMER;
+    }
+    if (alarm_touch_touchalarm_woke_us_up()) {
+        return NRF_SLEEP_WAKEUP_TOUCHPAD;
+    }
+    if (reset_reason_saved & NRF_POWER_RESETREAS_RESETPIN_MASK) {
+        return NRF_SLEEP_WAKEUP_RESETPIN;
+    }
+    else if (reset_reason_saved & NRF_POWER_RESETREAS_OFF_MASK) {
+        return NRF_SLEEP_WAKEUP_GPIO;
+    }
+    else if (reset_reason_saved & NRF_POWER_RESETREAS_VBUS_MASK) {
+        return NRF_SLEEP_WAKEUP_VBUS;
+    }
+    return NRF_SLEEP_WAKEUP_UNDEFINED;
+}
+
+#ifdef NRF_DEBUG_PRINT
+static const char* cause_str[] = {
+  "UNDEFINED",
+  "GPIO",
+  "TIMER",
+  "TOUCHPAD",
+  "VBUS",
+  "RESETPIN",
+};
+void print_wakeup_cause(nrf_sleep_source_t cause) {
+    if (cause >= 0 && cause < NRF_SLEEP_WAKEUP_ZZZ) {
+        dbg_printf("wakeup cause = NRF_SLEEP_WAKEUP_%s\r\n",
+		   cause_str[(int)cause]);
+    }
+}
+#endif
+
+bool alarm_woken_from_sleep(void) {
+   nrf_sleep_source_t cause = _get_wakeup_cause();
+#ifdef NRF_DEBUG_PRINT
+   if (cause != NRF_SLEEP_WAKEUP_UNDEFINED) {
+       //print_wakeup_cause(cause);
+   }
+#endif
+   return (cause == NRF_SLEEP_WAKEUP_GPIO || cause == NRF_SLEEP_WAKEUP_TIMER
+	   || cause == NRF_SLEEP_WAKEUP_TOUCHPAD);
+}
+
+nrf_sleep_source_t alarm_woken_from_sleep_2(void) {
```

Comment:
Yes it is a superset but the original version is also used in main.c, so I leave both. I admit the name ".._2" is not so good.

Reply:
Is there a reason the original cannot use this expanded implementation? If they need to be separate, maybe it could be `idle_woken_from_sleep`, or something.